x-pack/plugin/geoip-enterprise-downloader/src/main/java/org/elasticsearch/xpack/geoip/EnterpriseGeoIpDownloaderLicenseListener.java

```diff
@@ -0,0 +1,154 @@
+        maybeUpdateTaskState(clusterService.state());
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        hasIngestGeoIpMetadata = hasIngestGeoIpMetadata(event.state());
+        final boolean ingestGeoIpCustomMetaChangedInEvent = event.metadataChanged()
+            && event.changedCustomMetadataSet().contains(INGEST_GEOIP_CUSTOM_METADATA_TYPE);
+        final boolean masterNodeChanged = Objects.equals(
+            event.state().nodes().getMasterNode(),
+            event.previousState().nodes().getMasterNode()
+        ) == false;
+        /*
+         * We don't want to potentially start the task on every cluster state change, so only maybeUpdateTaskState if this cluster change
+         * event involved the addition of custom geoip metadata OR a master node change
+         */
+        if (ingestGeoIpCustomMetaChangedInEvent || (masterNodeChanged && hasIngestGeoIpMetadata)) {
+            maybeUpdateTaskState(event.state());
+        }
+    }
+
+    private void maybeUpdateTaskState(ClusterState state) {
+        if (onMasterNode(state)) {
+            if (licenseIsValid) {
+                if (hasIngestGeoIpMetadata) {
+                    ensureTaskStarted();
+                }
+            } else {
+                ensureTaskStopped();
+            }
+        }
+    }
+
+    private void ensureTaskStarted() {
+        assert licenseIsValid : "Task should never be started without valid license";
+        persistentTasksService.sendStartRequest(
+            ENTERPRISE_GEOIP_DOWNLOADER,
+            ENTERPRISE_GEOIP_DOWNLOADER,
+            new EnterpriseGeoIpTaskParams(),
+            MASTER_TIMEOUT,
+            ActionListener.wrap(r -> logger.debug("Started enterprise geoip downloader task"), e -> {
+                Throwable t = e instanceof RemoteTransportException ? ExceptionsHelper.unwrapCause(e) : e;
+                if (t instanceof ResourceAlreadyExistsException == false) {
+                    logger.error("failed to create enterprise geoip downloader task", e);
+                }
+            })
+        );
+    }
+
+    private void ensureTaskStopped() {
+        ActionListener<PersistentTasksCustomMetadata.PersistentTask<?>> listener = ActionListener.wrap(
+            r -> logger.debug("Stopped enterprise geoip downloader task"),
+            e -> {
+                Throwable t = e instanceof RemoteTransportException ? ExceptionsHelper.unwrapCause(e) : e;
+                if (t instanceof ResourceNotFoundException == false) {
+                    logger.error("failed to remove enterprise geoip downloader task", e);
+                }
+            }
+        );
+        persistentTasksService.sendRemoveRequest(ENTERPRISE_GEOIP_DOWNLOADER, MASTER_TIMEOUT, listener);
+    }
+
+    private boolean hasIngestGeoIpMetadata(ClusterState state) {
+        return state.metadata().custom(INGEST_GEOIP_CUSTOM_METADATA_TYPE) != null;
+    }
```

Comment:
This is another single-use single-line helper, I think it's clearer to collapse it so that a reader doesn't have to jump to a different place to see the implementation?